### PR TITLE
[FEATURE] Discover endpoints through OPTIONS calls

### PR DESCRIPTION
**Pull request by e0ipso**
Respond with information about the data using the OPTIONS calls. The
information is managed at the ResourceField level with the
FieldPublicInfo objects.

### CODE REVIEW

**Comment by e0ipso** [comment]
This PR was already big enough as is. I added the following follow ups:
- Return appropriate headers for an OPTIONS call (#633).
- Bring back discovery testing (#634).

File: src/Plugin/formatter/FormatterJsonApi.php [hunk 1]
```diff
@@ -101,7 +101,12 @@ protected function extractFieldValues($data, &$included, array $parents = array(
       }
       if ($resource = $this->getResource()) {
         $output['type'] = $resource->getResourceMachineName();
-        $output['id'] = (string) $data->getIdField()->value($data->getInterpreter());
+        $resource_id = $data->getIdField()->value($data->getInterpreter());
+        if (!is_array($resource_id)) {
+          // In some situations when making an OPTIONS call the $resource_id
+          // returns the array of discovery information instead of a real value.
```
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJsonApi.php:107]: this sounds weird. Are we masking a bug here?

File: src/Plugin/resource/DataProvider/DataProvider.php [hunk 2]
```diff
@@ -244,6 +246,27 @@ public function index() {
   /**
    * {@inheritdoc}
    */
+  public function discover($path = NULL) {
+    // Alter the field definition by adding a callback to get the auto
+    // discover information in render time.
+    foreach ($this->fieldDefinitions as $public_field_name => $resouce_field) {
+      // If the given field does not have discovery information, provide the
+      // empty one instead of an error.
+      $callable = array('\Drupal\restful\Plugin\resource\Field\ResourceFieldBase::emptyDiscoveryInfo', array($public_field_name));
```
Comment (amitaibu) [inline src/Plugin/resource/DataProvider/DataProvider.php]: I'd move this to an `else` below - just for readability

File: src/Plugin/resource/DataProvider/DataProviderInterface.php [hunk 3]
```diff
@@ -12,6 +12,18 @@
...
+  /**
+   * Return the discovery information for the given entity.
+   *
+   * @param string $path
+   *   The request path.
+   *
+   * @return array
+   *   An array of data for the thing being discovered.
+   */
+  public function discover($path = NULL);
```
Comment (amitaibu) [inline src/Plugin/resource/DataProvider/DataProviderInterface.php:25]: `getDiscoveryInfo`. `discover` is more direct, but doesn't sound as returning the info. not sure.

File: src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+      'allowed_values' => NULL,
+    ),
+  );
+
+  /**
+   * Sections for the field information.
+   *
+   * @var array[]
+   */
+  protected $sections = array();
```
Comment (amitaibu) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: you mean the info | data | form_element? maybe call it `category`?

File: src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    elseif ($section_name == 'form_element') {
+      $this->validateFormElement($section_info);
+    }
+  }
+
+  /**
+   * Processes the provided data for the section.
+   *
+   * @param string $section_name
+   *   The name of the sections. By default RESTful suports 'info',
```
Comment (amitaibu) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: suports => supports

File: src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+   *
+   * @param array $section_info
+   *   The structured array with the section information.
+   *
+   * @throws ServerConfigurationException
+   *   If the field does not pass validation.
+   */
+  protected function validateInfo(array $section_info) {
+    if (empty($section_info['label'])) {
+      throw new ServerConfigurationException(sprintf('The basic information is not valid for this field: %s.', $this->fieldName));
```
Comment (amitaibu) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: maybe: `The label information is missing for this ...`

File: src/Plugin/resource/Field/ResourceFieldEntity.php [hunk 7]
```diff
@@ -410,6 +414,11 @@ public function access($op, DataInterpreterInterface $interpreter) {
    * @throws ServerConfigurationException
    */
   protected function propertyWrapper(DataInterpreterInterface $interpreter) {
+    // This is the first method that gets called for all fields after loading
+    // the entity. We'll use that opportunity to set the actual bundle of the
+    // field.
+    $this->setBundle($interpreter->getWrapper()->getBundle());
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:420]: why not do this in the constructor?

File: src/Plugin/resource/Field/ResourceFieldEntity.php [hunk 8]
```diff
@@ -113,9 +115,11 @@ public function __construct(array $field, RequestInterface $request) {
     $this->wrapperMethodOnEntity = isset($field['wrapper_method_on_entity']) ? $field['wrapper_method_on_entity'] : $this->wrapperMethodOnEntity;
     $this->column = isset($field['column']) ? $field['column'] : $this->column;
     $this->imageStyles = isset($field['image_styles']) ? $field['image_styles'] : $this->imageStyles;
-    if (!empty($field['bundles'])) {
-      $this->bundles = $field['bundles'];
+    if (!empty($field['bundle'])) {
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:118]: what's this part - I don't remember?

File: src/Plugin/resource/Field/ResourceFieldEntity.php [hunk 9]
```diff
@@ -782,15 +772,26 @@ protected function entityTypeWrapper() {
...
+  public function setBundle($bundle) {
+    // Do not do pointless work if not needed.
+    if (!empty($this->bundle) && $this->bundle == $bundle) {
+      return;
+    }
+    $this->bundle = $bundle;
+
+    // If this is an options call, then introspect Entity API to add more data
+    // to the public field information.
+    if ($this->getRequest()->getMethod() != RequestInterface::METHOD_OPTIONS) {
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntity.php]: I'd change the logic to `==` so the above comment would make more sense - just a minor readability thing.

File: src/Plugin/resource/Field/ResourceFieldEntity.php [hunk 10]
```diff
@@ -1020,6 +1021,16 @@ public function isComputed() {
   /**
    * {@inheritdoc}
    */
+  public function autoDiscovery() {
+    if (method_exists($this->decorated, 'autoDiscovery')) {
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:1024]: what's the use case for _not_ having autoDiscovery?

File: src/Plugin/resource/Field/ResourceFieldEntityText.php [hunk 11]
```diff
@@ -18,23 +18,9 @@ class ResourceFieldEntityText extends ResourceFieldEntity implements ResourceFie
   public function preprocess($value) {
     // Text field. Check if field has an input format.
 
-    // This is a big leap of faith, but here we go... In order to have more
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntityText.php:21]: so we're back to single bundle! this sounds much better to me :)

File: src/Plugin/resource/DataProvider/DataProviderInterface.php (hunk 3 above)
Comment (amitaibu) [inline src/Plugin/resource/DataProvider/DataProviderInterface.php:25]: `discover` is fine - it's in the same spirit of `replace`, `update`, ...

**Comment by amitaibu** [comment]
Mostly question/ minor naming issues.

**Comment by amitaibu** [comment]
RESTful is pretty cool ;)

File: src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php (hunk 4 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: I'm OK either way.

:+1:

Comment (amitaibu) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: so `category`sounds better I think. section for me has a visual connotation, rather then data.

File: src/Plugin/resource/DataProvider/DataProvider.php (hunk 2 above)
Comment (e0ipso) [inline src/Plugin/resource/DataProvider/DataProvider.php]: :+1:

File: src/Plugin/resource/DataProvider/DataProviderInterface.php (hunk 3 above)
Comment (e0ipso) [inline src/Plugin/resource/DataProvider/DataProviderInterface.php:25]: :+1:

File: src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php (hunk 5 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/PublicFieldInfo/PublicFieldInfoBase.php]: :+1:

File: src/Plugin/resource/Field/ResourceFieldEntity.php (hunk 8 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:118]: This was not used until now, that's why we didn't realize that it made no sense to have multiple bundles in a field instance for a loaded entity.

Having a single bundle simplifies things greatly.

File: src/Plugin/resource/Field/ResourceFieldEntity.php (hunk 7 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:420]: This gets constructed with the `Resource` when the request comes in. That means that the entity is not loaded yet, which in turn means that we don't know the bundle of the entity.

After loading the entity this is the first method that gets called for all entity-based resource fields that implies an entity being loaded and has access to it through the `$interpreter`.

File: src/Plugin/resource/Field/ResourceFieldEntity.php (hunk 9 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntity.php]: yeah, I had a big chunk of code before so I wanted to save an indentation level. The I refactored it into a function.

File: src/Plugin/resource/Field/ResourceFieldEntity.php (hunk 10 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:1024]: There is none, but I don't want to enforce it for new resource fields that implementors (and related modules) can create.

File: src/Plugin/resource/Field/ResourceFieldEntityText.php (hunk 11 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntityText.php:21]: Yeah, but only for the resource fields not for the resources ;-)

**Comment by e0ipso** [comment]
@amitaibu feedback implemented.

**Comment by e0ipso** [comment]
@amitaibu tests are green. Travis CI had a hiccup.

**Comment by amitaibu** [comment]
:+1:

**Pull request merged**